### PR TITLE
Dennis/gps2

### PR DESCRIPTION
**Pull request by dennisfedorko**
Halfway point for GPS integration. 

GPS.swift should process all GPS reading/writing events for controllers through adding observers as delegates. Instead of using notifications, I created a multi-delegate interface so multiple controllers could listen for GPS updates if they so choose.

Backend component has been implemented/tested

Need to create driver login for logging bus location via DriverViewController which is currently displayed if we tap the tab bar navigation button 5 times in a row. (currently a blank screen)

Need to create marker/image for shuttle bus to fully test on map view.

We also need to consider edge cases for how to react if user is not logging correctly or some other error happens

### CODE REVIEW

File: big-red-shuttle/AppDelegate.swift [hunk 1]
```diff
@@ -60,31 +60,55 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+//        }, failure: { (json: JSON?) in
+//            print("here")
+//        })
+////
+//        api().getLocation(success: { (json: JSON) in
+//            print("here")
+//
+//        }, failure: { (json: JSON) in
+//            print("here")
+//        })
```
Comment (anniexcheng) [inline big-red-shuttle/AppDelegate.swift:90]: Do we still this commented out code for future testing?

File: big-red-shuttle/AppDelegate.swift [hunk 2]
```diff
@@ -60,31 +60,55 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
     }
+    
+    var hiddenTabCounter = 0
+    
+    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
+        if tabBarController.selectedIndex == 0 {
+            hiddenTabCounter += 1
+        } else {
+            hiddenTabCounter = 0
+        }
```
Comment (anniexcheng) [inline big-red-shuttle/AppDelegate.swift]: condense lines 101-105:
`hiddenTabCounter = tabBarController.selectedIndex == 0 ? hiddenTabCounter + 1 : 0`

File: big-red-shuttle/GPS.swift [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        super.init()
+    }
+    
+    func addObserver(observer: ShuttleBusGPSDelegate) {
+        weak var weakObserver = observer
+        observers.append(weakObserver)
+    }
+    
+    func removeObserver(observer: ShuttleBusGPSDelegate) {
+        weak var weakObserver = observer
```
Comment (anniexcheng) [inline big-red-shuttle/GPS.swift:33]: should prob use `let` since `weakObserver` is never changed

File: big-red-shuttle/AppDelegate.swift (hunk 1 above)
Comment (dennisfedorko) [reply big-red-shuttle/AppDelegate.swift:90]: Ya why not

File: big-red-shuttle/GPS.swift [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    func logShuttleLocation() {
+        if !isLoggingShuttleLocation { return }
+        
+        api().logLocation(success: { (json: JSON?) in
+            Timer.scheduledTimer(timeInterval: self.fetchUpdateFrequency,
+                                 target: self,
+                                 selector: #selector(self.fetchShuttleLocation),
+                                 userInfo: nil,
+                                 repeats: false)
+        }, failure: { (json: JSON?) in
```
Comment (anniexcheng) [inline big-red-shuttle/GPS.swift:64]: lines 56 and 64: use `_` instead of `(json: JSON?)` since `json` is never used

File: big-red-shuttle/GPS.swift [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    
+    func stopFetchingShuttleLocation() {
+        isFetchingShuttleLocation = false
+        notifyObserversGPSStoppedFetching()
+    }
+    
+    @objc func fetchShuttleLocation() {
+        if !isFetchingShuttleLocation { return }
+        
+        api().getLocation(success: { (json: JSON?) in
```
Comment (anniexcheng) [inline big-red-shuttle/GPS.swift:85]: `_` instead of `(json: JSON?)`

File: big-red-shuttle/GPS.swift (hunk 3 above)
Comment (dennisfedorko) [reply big-red-shuttle/GPS.swift:33]: can't have weak let since by definition weak variables can change to nil

File: big-red-shuttle/GPS.swift [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                Timer.scheduledTimer(timeInterval: self.fetchUpdateFrequency,
+                                     target: self,
+                                     selector: #selector(self.fetchShuttleLocation),
+                                     userInfo: nil,
+                                     repeats: false)
+                
+            } else {
+                self.stopFetchingShuttleLocation()
+            }
+        }, failure: { (json: JSON?) in
```
Comment (anniexcheng) [inline big-red-shuttle/GPS.swift:97]: same as line 85

File: big-red-shuttle/GPS.swift (hunk 4 above)
Comment (dennisfedorko) [reply big-red-shuttle/GPS.swift:64]: I actually don't like using _ because then devs can't figure out what that variable is at a quick glance. Similar to why I leave the types in the block declarations

File: big-red-shuttle/GPS.swift (hunk 6 above)
Comment (dennisfedorko) [reply big-red-shuttle/GPS.swift:97]: negative

File: big-red-shuttle/GPS.swift (hunk 5 above)
Comment (dennisfedorko) [reply big-red-shuttle/GPS.swift:85]: check above comment ^^

File: big-red-shuttle/GPS.swift (hunk 3 above)
Comment (anniexcheng) [reply big-red-shuttle/GPS.swift:33]: ah gotcha

File: big-red-shuttle/GPS.swift (hunk 4 above)
Comment (anniexcheng) [reply big-red-shuttle/GPS.swift:64]: alright, up to you

File: big-red-shuttle/Network.swift [hunk 7]
```diff
@@ -34,45 +34,44 @@ class System {
...
+            failure?(nil)
             return
         }
         
         let parameters = ["uid": uid, "key": key]
         
-        request(endpoint: "/register", parameters: parameters, method: .post, encoding: JSONEncoding.default) { (response: JSON) in
-            completion?()
-        }
+        request(endpoint: "/register", parameters: parameters, method: .post, encoding: JSONEncoding.default, success: success, failure: failure)
```
Comment (anniexcheng) [inline big-red-shuttle/Network.swift:50]: believe you don't need `JSONEncoding`

File: big-red-shuttle/Network.swift [hunk 8]
```diff
@@ -81,21 +80,32 @@ class API {
...
         
-        request(endpoint: "/latest", parameters: parameters, method: .get, encoding: URLEncoding.default) { (response: JSON) in
-            let latitude = CLLocationDegrees(response["latitude"].floatValue)
-            let longitude = CLLocationDegrees(response["longitude"].floatValue)
-            let coordinate = Coordinate(latitude: latitude, longitude: longitude)
-            let timestamp = response["date"].stringValue
-            let dateFormatter = DateFormatter()
-            dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSSSSS"
-            let date = dateFormatter.date(from: timestamp)
+        request(endpoint: "/latest", parameters: parameters, method: .get, encoding: URLEncoding.default, success: { (json: JSON?) in
```
Comment (anniexcheng) [inline big-red-shuttle/Network.swift:83]: same as line 50

**Comment by anniexcheng** [comment]
alright, left some comments, otherwise looks good!

File: big-red-shuttle/Network.swift (hunk 8 above)
Comment (dennisfedorko) [reply big-red-shuttle/Network.swift:83]: it's necessary with alamofire

File: big-red-shuttle/Network.swift (hunk 7 above)
Comment (dennisfedorko) [reply big-red-shuttle/Network.swift:50]: its necessary since you also have URLEncoding.default

File: big-red-shuttle/Network.swift (hunk 8 above)
Comment (anniexcheng) [reply big-red-shuttle/Network.swift:83]: alright, didn't try compiling so wasn't sure
